File: backend/src/main/java/com/decmoe47/todo/controller/TodoListController.java
package com.decmoe47.todo.controller;

import com.decmoe47.todo.model.dto.TodoListAddDTO;
import com.decmoe47.todo.model.dto.TodoListDeleteDTO;
import com.decmoe47.todo.model.dto.TodoListUpdateDTO;
import com.decmoe47.todo.model.vo.R;
import com.decmoe47.todo.model.vo.TodoListVO;
import com.decmoe47.todo.service.TodoListService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/todoLists")
public class TodoListController {

    private final TodoListService todoListService;

    @GetMapping("/custom")
    public R<List<TodoListVO>> getCustomTodoLists() {
        return R.ok(todoListService.getCustomTodoLists());
    }

    @PostMapping("/add")
    public R<TodoListVO> addTodoList(@RequestBody TodoListAddDTO todoListAddDTO) {
        return R.ok(todoListService.addTodoList(todoListAddDTO));
    }

    @PostMapping("/update")
    @PreAuthorize("@accessCheckService.ownsTodoListAccess(authentication.principal.id, #todoListUpdateDTO.id)")
    public R<TodoListVO> updateTodoList(@RequestBody TodoListUpdateDTO todoListUpdateDTO) {
        return R.ok(todoListService.updateTodoList(todoListUpdateDTO));
    }

    @PostMapping("/delete")
    @PreAuthorize("@accessCheckService.ownsTodoListAccess(authentication.principal.id, #todoListDeleteDTO.id)")
    public R<Object> deleteTodoList(@RequestBody TodoListDeleteDTO todoListDeleteDTO) {
        todoListService.deleteTodoList(todoListDeleteDTO);
        return R.ok();
    }
}

File: backend/src/main/kotlin/com/decmoe47/todo/controller/TodoListController.kt
package com.decmoe47.todo.controller

import com.decmoe47.todo.model.request.TodoListAddRequest
import com.decmoe47.todo.model.request.TodoListDeleteRequest
import com.decmoe47.todo.model.request.TodoListUpdateRequest
import com.decmoe47.todo.model.response.Response
import com.decmoe47.todo.model.response.TodoListResponse
import com.decmoe47.todo.service.TodoListService
import com.decmoe47.todo.util.R
import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/todoLists")
class TodoListController(private val todoListService: TodoListService) {
    @GetMapping("/all")
    fun getAll(): ResponseEntity<Response<List<TodoListResponse>>> =
        R.ok(todoListService.getAll())

    @PostMapping("/add")
    fun add(@Valid @RequestBody request: TodoListAddRequest): ResponseEntity<Response<TodoListResponse>> =
        R.ok(todoListService.add(request))

    @PostMapping("/update")
    @PreAuthorize("@accessCheckService.ownsTodoListAccess(authentication.principal.id, #request.id)")
    fun update(@Valid @RequestBody request: TodoListUpdateRequest): ResponseEntity<Response<TodoListResponse>> =
        R.ok(todoListService.update(request))

    @PostMapping("/delete")
    @PreAuthorize("@accessCheckService.ownsTodoListAccess(authentication.principal.id, #request.id)")
    fun delete(@Valid @RequestBody request: TodoListDeleteRequest): ResponseEntity<Response<Unit>> {
        todoListService.delete(request)
        return R.ok()
    }
}
